### ISSUE TITLE
Guard-livereload seems to have suddenly failed

### ISSUE THREAD

**Issue by crispen-smith** [guard/guard-livereload#130]
2 days ago guard was working beautifully,  as of the last couple of days however, no dice.

I'v rebuilt Ruby, cleaned all gems, cleaned Node (this is a node server that I'm running against), I've built a fresh version of bundler, rebundled guard and guard-reload  and still.... nada

It connects to the Chrome extension just fine but then it seems to hang until I quit it.

I am running Sass-watch, nodemon concurrently but I was running these with successful starts as well.  

Guardfile is:

guard 'livereload' do
  watch(%r{.+\.(css|js|ejs|html)})
end

**Comment by e2** [comment]
1. Run guard with the '-d' option and show the output
2. you may want to try and run *just* the livereload plugin (e.g. guard -P `livereload`) - this is because livereload uses EventMachine and maybe something out there is using fork()
3. check out https://github.com/guard/guard/wiki/Understanding-Guard to make sure all the watches are working correctly, that the events are triggered - there are too many things that can go wrong in that area
4. if you can reproduce the issue in a separate repo, I can take a look

**Comment by crispen-smith** [comment]
Here's the debug output.


Crispens-MacBook-Pro:site crispensmith$ bundle exec guard -d
19:55:53 - DEBUG - Notiffany: gntp not available (Please add "gem 'ruby_gntp'" to your Gemfile and run your app with "bundle exec".).
19:55:53 - DEBUG - Notiffany: growl not available (Please add "gem 'growl'" to your Gemfile and run your app with "bundle exec".).
19:55:53 - DEBUG - Notiffany: terminal_notifier not available (Please add "gem 'terminal-notifier-guard'" to your Gemfile and run your app with "bundle exec".).
19:55:53 - DEBUG - Notiffany: libnotify not available (Unsupported platform "darwin14.1.0").
19:55:53 - DEBUG - Notiffany: notifysend not available (Unsupported platform "darwin14.1.0").
19:55:53 - DEBUG - Notiffany: notifu not available (Unsupported platform "darwin14.1.0").
19:55:53 - DEBUG - Command execution: emacsclient --eval '1'
19:55:53 - DEBUG - Notiffany: emacs not available (Emacs client failed).
19:55:53 - DEBUG - Notiffany: tmux not available (:tmux notifier is only available inside a TMux session.).
19:55:53 - DEBUG - Notiffany: file not available (No :path option given).
19:55:53 - DEBUG - Notiffany is using TerminalTitle to send notifications.
19:55:53 - DEBUG - Command execution: hash stty
19:55:53 - DEBUG - Guard starts all plugins
19:55:53 - DEBUG - Hook :start_begin executed for Guard::LiveReload
19:55:53 - DEBUG - Hook :start_end executed for Guard::LiveReload
19:55:53 - INFO - LiveReload is waiting for a browser to connect.
19:55:53 - INFO - Guard is now watching at '/Users/crispensmith/Documents/cinchedstore/site'
19:55:53 - DEBUG - Start interactor
19:55:53 - DEBUG - Command execution: stty -g 2>/dev/null
[1] guard(main)> 19:56:38 - INFO - Browser connected.
19:56:42 - DEBUG - Serving file /usr/local/lib/ruby/gems/2.2.0/gems/guard-livereload-2.4.0/js/livereload.js
19:56:42 - DEBUG - Serving file /usr/local/lib/ruby/gems/2.2.0/gems/guard-livereload-2.4.0/js/livereload.js


Once the browser connected I did save a file, but there's no evidence of this in the debugging output.   Not sure if there should be.

**Comment by e2** [comment]
That's why you should go through the: https://github.com/guard/guard/wiki/Understanding-Guard
It looks a bit long, but ultimately it will save you a ton of time if you customize projects using Guard in anyway.

For example, the debug output above lacks this kind of output: https://github.com/guard/guard/wiki/Understanding-Guard#is-the-file-passed-to-the-plugin

Which means your watcher is wrong (I didn't notice at first).

Check this section to work out the correct matcher for your project:

https://github.com/guard/guard/wiki/Understanding-Guard#debugging-the-matchers-in-guard

(You'll want to watch just the output files, mostly).

Let me know if there's anything I can improve in Guard or the docs to help work out things like this faster.

**Comment by crispen-smith** [comment]
Hey E2, I'm a little confused as I thought I had written back to this on Friday...  but nonetheless.

I'm not entirely sure the watcher is wrong, but more importantly; I think the issue may be deeper than I originally expected as I've opened this ticket (https://github.com/sass/sass/issues/1631) against sass and I think it's related.

For the question around my watchers, the  logic I had been using was: 1) .JS for my engine, 2) .EJS for my templates 3) .CSS for my compiled css

I did a test with *.* with just an echo, and didn't see any matches at all, so I wonder if it's something deeper.

**Comment by e2** [comment]
If you don't see: `Hook :run_on_modifications_begin executed for Guard::Livereload`, it means the files you changed didn't match the the watch expression.

If you can, create a new repository so I can reproduce the error.

Or, it could be a known issue (e.g. a Dropbox folder, or something). 

Check this for possible issues: https://github.com/guard/listen/wiki (both links in Troubleshooting section)

(Polling has to work, or there's something wrong - check out the available Listen options you can pass to Guard here: https://github.com/guard/guard/wiki/Command-line-options-for-Guard#listen-options - especially try polling)

**Comment by e2** [comment]
BTW, ideally, your regexp should be something like: `%r{^.+\.(css|js|ejs|html)$})`.

Either way try the Listen troubleshooting to rule out problems with events not getting fired (upload Listen debug output in a Gist if you're in doubt).

**Comment by crispen-smith** [comment]
Okay... so I've read through and I think I'm more confused now then ever.

To be truly safe and verbose, I've mapped my Guardfile to:

watch(/(.*)/) { |m| Guard::UI.puts "Unknown file: #{m[1]}"; nil }

guard :shell do
  watch /(.*)/ do |m|
    `echo #{m[0]}`
  end
end

With the expectation that this should at least provide some useful output from the debugger.  However; using 2 terminals I've got the following output

WINDOW 1:
^CCrispens-MacBook-Pro:cinchedstore crispensmith$ bundle exec guard -d
21:30:18 - DEBUG - Notiffany: gntp not available (Please add "gem 'ruby_gntp'" to your Gemfile and run your app with "bundle exec".).
21:30:18 - DEBUG - Notiffany: growl not available (Please add "gem 'growl'" to your Gemfile and run your app with "bundle exec".).
21:30:18 - DEBUG - Notiffany: terminal_notifier not available (Please add "gem 'terminal-notifier-guard'" to your Gemfile and run your app with "bundle exec".).
21:30:18 - DEBUG - Notiffany: libnotify not available (Unsupported platform "darwin14.1.0").
21:30:18 - DEBUG - Notiffany: notifysend not available (Unsupported platform "darwin14.1.0").
21:30:18 - DEBUG - Notiffany: notifu not available (Unsupported platform "darwin14.1.0").
21:30:18 - DEBUG - Command execution: emacsclient --eval '1'
21:30:18 - DEBUG - Notiffany: emacs not available (Emacs client failed).
21:30:18 - DEBUG - Notiffany: tmux not available (:tmux notifier is only available inside a TMux session.).
21:30:18 - DEBUG - Notiffany: file not available (No :path option given).
21:30:18 - DEBUG - Notiffany is using TerminalTitle to send notifications.
21:30:19 - DEBUG - Command execution: hash stty
21:30:19 - DEBUG - Guard starts all plugins
21:30:19 - DEBUG - Hook :start_begin executed for Guard::Shell
21:30:19 - DEBUG - Hook :start_end executed for Guard::Shell
21:30:19 - INFO - Guard is now watching at '/Users/crispensmith/Documents/cinchedstore'
21:30:19 - DEBUG - Start interactor
21:30:19 - DEBUG - Command execution: stty -g 2>/dev/null
[1] guard(main)> 

Window 2: 
Last login: Sun Feb  8 20:26:36 on ttys003
Crispens-MacBook-Pro:cinchedstore crispensmith$ pwd
/users/crispensmith/documents/cinchedstore
Crispens-MacBook-Pro:cinchedstore crispensmith$ touch test.css
Crispens-MacBook-Pro:cinchedstore crispensmith$ touch test2.css
Crispens-MacBook-Pro:cinchedstore crispensmith$ touch test3.css
Crispens-MacBook-Pro:cinchedstore crispensmith$ 

My understanding is that I should see some sign of these touches.

I've also tried to suss my way through debugging Listen.  That process did identify for me that I needed to add the celluloid-oi gem; which I've done.

However, as I'm quite far removed from Ruby I'm not sure how to best implement the instruction:
Use the LISTEN_GEM_DEBUGGING set to 2 for additional info.

**Comment by e2** [comment]
It's not Ruby specific, it's an environment variable that you can set e.g.

```bash
LISTEN_GEM_DEBUGGING=2 bundle exec guard -d
```

Or you can set it once per shell session, e.g.

```bash
export LISTEN_GEM_DEBUGGING=2
bundle exec guard -d
```

Just focus on this, because if you're not getting any notifications, `LISTEN_GEM_DEBUGGING=2` will probably give the best insights.

**Comment by crispen-smith** [comment]
And that's exactly what I wasn't sure of as far as the flags go;  was there a config file, was it just a CLI call... too many options.

I'll go ahead with this and see what I can come up.

**Comment by crispen-smith** [comment]
Here's a trace, with `touch test4.css` executed during the first operation, `touch test5.css` during the second (listen only).

Crispens-MacBook-Pro:cinchedstore crispensmith$ export LISTEN_GEM_DEBUGGING=2
Crispens-MacBook-Pro:cinchedstore crispensmith$ bundle exec guard -d
I, [2015-02-09T18:32:26.943213 #6464]  INFO -- : Celluloid loglevel set to: 0
18:32:26 - DEBUG - Notiffany: gntp not available (Please add "gem 'ruby_gntp'" to your Gemfile and run your app with "bundle exec".).
18:32:26 - DEBUG - Notiffany: growl not available (Please add "gem 'growl'" to your Gemfile and run your app with "bundle exec".).
18:32:26 - DEBUG - Notiffany: terminal_notifier not available (Please add "gem 'terminal-notifier-guard'" to your Gemfile and run your app with "bundle exec".).
18:32:26 - DEBUG - Notiffany: libnotify not available (Unsupported platform "darwin14.1.0").
18:32:26 - DEBUG - Notiffany: notifysend not available (Unsupported platform "darwin14.1.0").
18:32:26 - DEBUG - Notiffany: notifu not available (Unsupported platform "darwin14.1.0").
18:32:26 - DEBUG - Command execution: emacsclient --eval '1'
18:32:27 - DEBUG - Notiffany: emacs not available (Emacs client failed).
18:32:27 - DEBUG - Notiffany: tmux not available (:tmux notifier is only available inside a TMux session.).
18:32:27 - DEBUG - Notiffany: file not available (No :path option given).
18:32:27 - DEBUG - Notiffany is using TerminalTitle to send notifications.
18:32:27 - DEBUG - Command execution: hash stty
18:32:27 - DEBUG - Guard starts all plugins
18:32:27 - DEBUG - Hook :start_begin executed for Guard::Shell
18:32:27 - DEBUG - Hook :start_end executed for Guard::Shell
D, [2015-02-09T18:32:27.538069 #6464] DEBUG -- : Adapter: considering TCP ...
D, [2015-02-09T18:32:27.538157 #6464] DEBUG -- : Adapter: considering polling ...
D, [2015-02-09T18:32:27.538182 #6464] DEBUG -- : Adapter: considering optimized backend...
I, [2015-02-09T18:32:30.494505 #6464]  INFO -- : Record.build(): 2.955162763595581 seconds
18:32:30 - INFO - Guard is now watching at '/Users/crispensmith/Documents/cinchedstore'
18:32:30 - DEBUG - Start interactor
18:32:30 - DEBUG - Command execution: stty -g 2>/dev/null
[1] guard(main)> quit
18:33:28 - DEBUG - Interactor was stopped or killed
18:33:28 - DEBUG - Command execution: stty  2>/dev/null
W, [2015-02-09T18:33:28.216476 #6464]  WARN -- : Terminating task: type=:finalizer, meta={:method_name=>:__shutdown__}, status=:receiving
	Celluloid::TaskFiber backtrace unavailable. Please try `Celluloid.task_class = Celluloid::TaskThread` if you need backtraces here.
18:33:28 - DEBUG - Guard stops all plugins
18:33:28 - DEBUG - Hook :stop_begin executed for Guard::Shell
18:33:28 - DEBUG - Hook :stop_end executed for Guard::Shell

18:33:28 - INFO - Bye bye...
Crispens-MacBook-Pro:cinchedstore crispensmith$ export LISTEN_GEM_DEBUGGING=2
Crispens-MacBook-Pro:cinchedstore crispensmith$ listen
Starting listen...
I, [2015-02-09T18:33:56.176611 #6483]  INFO -- : Celluloid loglevel set to: 0
D, [2015-02-09T18:33:56.364448 #6483] DEBUG -- : Broadcaster: starting tcp server: 127.0.0.1:4000
D, [2015-02-09T18:33:56.365206 #6483] DEBUG -- : Adapter: considering TCP ...
D, [2015-02-09T18:33:56.365244 #6483] DEBUG -- : Adapter: considering polling ...
D, [2015-02-09T18:33:56.365263 #6483] DEBUG -- : Adapter: considering optimized backend...
I, [2015-02-09T18:33:56.556520 #6483]  INFO -- : Record.build(): 0.18999910354614258 seconds

**Comment by e2** [comment]
Try Guard with the '-p' option (polling) - it should work, no matter what.

I think you're using an outdated version of Listen - it should log the version. Try:
```bash
bundle update listen
```

And just in case:

```bash
bundle update guard
bundle update celluloid
bundle update rb-fsevent
```

It looks like you're not getting any events, and either you're watching a special folder (Dropbox?) or there's something strange with your system (check maybe: http://feedback.livereload.com/knowledgebase/articles/86239).

You may want to go through the Known Issues in Listen - just to make sure: https://github.com/guard/listen/wiki/Troubleshooting#known-issues (I'm not a Mac user, so there's a limit to what I can suggest there).

On a Mac, the following gem is used for events: https://github.com/thibaudgg/rb-fsevent (you could check the issues for).

**Comment by crispen-smith** [comment]
I think you might have identified it with the rb-fsevent issue.  One of the proposed solutions is to move to a different directory and I've gotten a guardfile to post Interactor was stopped or killed  message on a touch.

So, let's go ahead and close this and chalk it up to that.  :)

**Comment by e2** [comment]
> "Interactor was stopped or killed"

That happens when a modified file is detected or when you quit Guard.

**Comment by crispen-smith** [comment]
Sorry if I wasn't clear... I was ware of this meaning of it from reading the docs you had me read.  So it was what I was looking for to call this solution a win.   Once I saw it I knew I was in good shape.

**Comment by e2** [comment]
Let me know if somehow having better docs would've helped you - mostly using Guard with the '-d' option and setting the environment variable `LISTEN_GEM_DEBUGGING=2` gives a good idea of what is wrong.

Feel free to update the Wiki - I'm in Guard too deep to make good decisions about what is helpful or not.

I opened a new issue: https://github.com/guard/guard/issues/741 (any feedback would be appreciated).

**Comment by crispen-smith** [comment]
I think the biggest issue in my particular case was the underlying problem with rb-fsevent.   It would seem that on a Mac rb-fsevent randomly *hates* camelCase directories.  Sadly, I think this means that all of my project names will have to follow the patter "projectname"  which is harder to read, but seems to guarantee functionality.

**Comment by e2** [comment]
@crispen-smith - wow! That's revealing! Are you sure it's a problem on the rb-fsevent level? It could be a bug in Listen (I'm not sure - can't test really since I only use Linux).

**Comment by crispen-smith** [comment]
@e2 , I'm not 100% certain, but I can say that it only came to light when i reviewed this: https://github.com/thibaudgg/rb-fsevent.

But... whether it's rb-fsevent or listen,  it was a frighteningly subtle bug in the sense that it just fails to see changes.  At one point I was debating replacing my machine.

Worst part was that it impacted EVERYTHING - Sass, Guard, forever.js (all of which probably call listen at some point in their lifecycle).   I definitely feel that it needs to be called out as part of the guard trouble-shooting, wherever the source of it may be.

**Comment by e2** [comment]
@crispen-smith - actually, the MacOS support has given me the most grief (especially since I don't own/use a Mac). The issue is mentioned in the Wiki (a link to livereload page with hints about filename case sensitivity). And there's a link to a (lengthy) issue about the bug in rb-fsevent (with solution here: https://github.com/thibaudgg/rb-fsevent/issues/10#issuecomment-20328752).

Did you try the rb-fsevent Ruby examples here: https://github.com/thibaudgg/rb-fsevent#singular-path (I'm guessing you did).

Polling worked, right? (Or did it?)

**Comment by e2** [comment]
Oh, and let me know if this can be "detected" - e.g. I could force current directory and watched directories to lowercase in Listen if that helps - or warn if non-lowercase names are used. And you have an example to reproduce the problem consistently, let me know (I would add a test case to Listen).

**Comment by crispen-smith** [comment]
I need to set aside some time in the next week or so to do some additional
testing to try and recreate it and run some tests/fixes.

Sadly, I won't have a window until early next week. :(

In principle... it would be great if the warning on namecasing were made
available.

**Comment by adispezio** [comment]
Ended up here from @nex3 's closed ticket.

Happened to me today. Sass apps watching my filesystem can't seem to recognize changes. Doesn't work for watching in Compass (command line), CodeKit, LiveReload, and Koala. Also seeing other apps no longer updating the filesystem. Sublime Text's folder view no longer updates.

Using the polling method of Compass works (--poll) but kills CPU.

I haven't updated rb-fsevent recently (to my knowledge), but I did recently upgrade osx to 10.10.2, might it be an issue with the fsevent api?
